### ISSUE TITLE
mlflow callback is not tirggered

### ISSUE THREAD

**Issue by jpointchoi** [ultralytics/ultralytics#8550]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

on_train_start, on_fit_epoc_end callbacks seems not be triggered. No print mesaages with below codes.

from ultralytics import YOLO, settings
import mlflow

# Update a setting
from ultralytics import settings
settings.update({'mlflow': False})
mlflow.end_run()

# Make sure MLflow is enabled in settings
#settings.update({'mlflow': True})
mlflow.set_tracking_uri("http://127.0.0.1:5000")
mlflow.set_experiment("mlflow_simple_new")

def on_train_start(trainer):
#    if mlflow:
    printf('in the on_train_start')

def on_fit_epoch_end(trainer):
#    if mlflow:
    printf('in the on_fit_epoc_end')
    metrics_dict = {f"{re.sub('[()]', '', k)}": float(v) for k, v in trainer.metrics.items()}
    mlflow.log_metrics(metrics=metrics_dict, step=trainer.epoch)

# Start an MLflow run
with mlflow.start_run():
    # Your training code here
    model = YOLO('yolov8n.pt')
    results = model.train(data='coco8.yaml', epochs=10, imgsz=640)

    model.add_callback("on_train_start",on_train_start)
    model.add_callback("on_fit_epoch_end",on_fit_epoch_end)

# Log metrics, parameters, or artifacts as needed
#    mlflow.log_param("epochs", 100)
#    mlflow.log_metrics({"mAP": results.metrics.map})  # Example metric

    # Don't forget to log your model at the end of training
    mlflow.pytorch.log_model(model, "model")
      # End the MLflow run
    mlflow.end_run()

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@jpointchoi hello! Thanks for reaching out and providing detailed information about the issue you're encountering. It seems like there might be a minor typo in your callback functions that could be causing the problem. The `printf` function should be `print`. Here's a corrected snippet of your callback functions:

```python
def on_train_start(trainer):
    print('in the on_train_start')

def on_fit_epoch_end(trainer):
    print('in the on_fit_epoch_end')
    metrics_dict = {f"{re.sub('[()]', '', k)}": float(v) for k, v in trainer.metrics.items()}
    mlflow.log_metrics(metrics=metrics_dict, step=trainer.epoch)
```

Additionally, ensure that the callbacks are added to the model before starting the training process. Here's how you can do it:

```python
model = YOLO('yolov8n.pt')
model.add_callback("on_train_start", on_train_start)
model.add_callback("on_fit_epoch_end", on_fit_epoch_end)
results = model.train(data='coco8.yaml', epochs=10, imgsz=640)
```

This adjustment ensures that the callbacks are properly registered with the model before the training begins. If you're still facing issues, please let us know! We're here to help. 😊
